src/scheduler/gc_work.rs

```diff
@@ -869,15 +871,113 @@ impl<E: ProcessEdgesWork, P: Plan<VM = E::VM> + PlanTraceObject<E::VM>> PlanScan
     }
 }
 
-impl<E: ProcessEdgesWork, P: Plan<VM = E::VM> + PlanTraceObject<E::VM>> ScanObjectsWork<E::VM>
-    for PlanScanObjects<E, P>
+impl<E: ProcessEdgesWork, P: Plan<VM = E::VM> + PlanTraceObject<E::VM>, const KIND: TraceKind>
+    ScanObjectsWork<E::VM> for PlanScanObjects<E, P, KIND>
 {
     type E = E;
 
     fn roots(&self) -> bool {
         self.roots
     }
 
+    fn do_work_common(
+        &self,
+        buffer: &[ObjectReference],
+        worker: &mut GCWorker<<Self::E as ProcessEdgesWork>::VM>,
+        mmtk: &'static MMTK<<Self::E as ProcessEdgesWork>::VM>,
+    ) {
+        let tls = worker.tls;
+
+        // If this is a root packet, the objects in this packet will have not been traced, yet.
+        //
+        // This step conceptually traces the edges from root slots to the objects they point to.
+        // However, VMs that deliver root objects instead of root edges are incapable of updating
+        // root slots.  Like processing an edge, we call `trace_object` on those objects, and
+        // assert the GC doesn't move those objects because we cannot store back to the slots.
+        //
+        // If this is a root packet, the `scanned_root_objects` variable will hold those root
+        // objects which are traced for the first time.
+        let scanned_root_objects = self.roots().then(|| {
+            // We create an instance of E to use its `trace_object` method and its object queue.
+            let mut process_edges_work =
+                <PlanProcessEdges<E::VM, P, TRACE_KIND_FAST>>::new(vec![], false, mmtk);
+
+            for object in buffer.iter().copied() {
+                let new_object = process_edges_work.trace_object(object);
```

Comment:
Here `trace_object` is a method of `process_edges_work`.  Generally speaking the `ProcessEdgesWork` (along with its `TraceKind` if relevant) is provided by the Plan (at `schedule_collection`)  But I think whether we pin the object is a different dimension.   These pinning roots needs to pinned regardless whether it is a normal collection or a defrag collection, or whether it is a marking trace or forwarding trace.  Probably it is better to add one more argument to `trace_object` itself, or add a different `trace_object_pin` method to `SFT` and `PlanTraceObject`.

Reply:
In other words, `TraceKind` is local to a trace (should be part of `TraceLocal` which disappeared when we ported MMTk to Rust), while whether to pin or not is specific to this `trace_object` call.